confmap/provider/eecprovider/e2e_test.go

```diff
@@ -0,0 +1,133 @@
+//go:build e2e
+
+package eecprovider // import "github.com/Dynatrace/dynatrace-otel-collector/confmap/provider/eecprovider/eecprovider"
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"os"
+	"os/exec"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestGetsConfig(t *testing.T) {
+	f, err := os.ReadFile("./testdata/otel-config.yaml")
+	require.NoError(t, err)
+	fs := &fileserver{config: f}
+	ts := httptest.NewServer(http.HandlerFunc(fs.HandleRequest))
+	cmd := exec.Command("../../../bin/dynatrace-otel-collector", fmt.Sprintf("--config=%s", configureProvider(t, ts.URL)))
```

Comment:
would it makes sense to mock the usage to the `dynatrace-otel-collector`? This way a bug in a different component can cause test failure here. I understand this is an e2e test, but maybe I would prefer to do e2e testing on a real k8s cluster (including manifests etc). WDYT?

Reply:
>would it makes sense to mock the usage to the `dynatrace-otel-collector`?

We could directly utilize the libraries the Collector uses to instantiate this component and test it that way, but I think this should be fairly reliable and saves us some effort. It also tests the e2e flow of building a Collector with this component (manifest.yaml -> OpenTelemetry Collector Builder -> Binary -> ...), which isn't strictly in the scope of these tests, but for the way we do "E2E" tests (tests that work with a built binary) I think it's okay. It will also alert us if any upstream changes break us in this process.

>This way a bug in a different component can cause test failure here.

For the most part we're only testing the pieces we are concerned about here, which are the modules that manage the Collector's configuration and internal lifecycle. We do depend a little bit on other components not to panic on startup, or on any extensions to actually respond to an HTTP request, but I don't expect that functionality will be fragile.

>I see these tests more as component tests than an e2e test. To do a proper e2e testing maybe I would prefer to do it on a real k8s cluster (including manifests etc). WDYT?

To me that would depend on what you're considering the system under test. In this case, I would say that the component is the confmap Provider and the system is the whole Collector + a configuration server, so this is E2E from the sense that we are testing the interaction between the Collector and configuration server with a full live Collector. I didn't want to bring any infrastructure setups into these tests since the functionality should behave the same regardless of whether we run in a VM, k8s, FaaS, etc.

That said, it's good to make sure we're covering all our bases. Given the above, do you think we're still missing something from these tests by not testing a Collector deployed to k8s?

Reply:
If the rest of the team is ok with this, I agree to leave it as it is

Reply:
In general using binaries in tests is not the best approach, although I agree it's the easier path. Still a proper E2E testing should happen on a cluster/container (since that's the expected env where the provider will run on), as with this approach we are not able to mock the environment behavior (such as network lag etc.) It should basically mock the user (machine) behavior. 

In my eyes we might want to re-evaluate these to component tests -> those should check the behavior of the component (provider) and prove that the component is responding to the system on different requests as expected, no matter from when the requests come from. Therefore we can mock the behavior of the used binary. Also it may be problematic to run them locally, since the test requires a pre-requisite of building the binary locally and moving it to the correct folder. Tests in general should be completely independent from other dependencies (except e2e). For example, missing of the binary will negatively affect the result of the test.

The decision is completely on you, I wanted to just point out the things that can bite us in the future.

Reply:
I think we're largely on the same page here, though I still think that deploying the Collector binary to test this code is too much setup for what essentially amounts to testing an HTTP request. If we were testing the whole Collector (env vars, resource limits, reading files, networking, etc.) I would be more inclined to suggest the same.

Right now the tests are gated behind a [e2e build flag](https://github.com/Dynatrace/dynatrace-otel-collector/pull/175/files#diff-873ab01c873087589fdce6adcee1d3dad3c77be37850bc0977a76c47dbcc5f93R1), so they will only be run if the flag is enabled, which we explicitly [do in the workflow](https://github.com/Dynatrace/dynatrace-otel-collector/pull/175/files#diff-a2de7550554c0198ac7f959c78a19ee0996921c98034411de47a7dd49b0c209bR174). Running `go test` locally won't have any dependencies on the binary unless whoever is running the tests also enables the flag.

>The decision is completely on you, I wanted to just point out the things that can bite us in the future.

Thanks for giving it some serious consideration. I'm trying to keep things simple for now because I expect this code will be temporary and seldom-modified while we have discussions in the OTel community to determine how to bring this functionality to the upstream Providers. If we end up needing to keep it long-term we can return to this and see what we can do to ensure our development approach is more robust.